### DISCUSSION TITLE
[Discussion] Ideas for Better Internal Links

### DISCUSSION THREAD

**Discussion by ThaddaeusWiedemer** [Zettlr/Zettlr#2410] [Ideas]
I'd like to discuss possible improvements to the handling and rendering of internal links, especially when linking to files using an ID. As internal links are essential to Zettlr, especially to a Zettelkasten workflow, I feel like the community could benefit greatly from improvements here. I have some ideas but no definitive answers and am also aware that some things depend on different Markdown styles (which I'm not really well versed in), so please feel free to share opinions.

I've also tried to compile related enhancements and feature requests here. Many things have been proposed and discussed in the path, but I feel like many topics are linked (no pun intended) and we could benefit from having a clearer vision what Zettlr's workflow with internal links could look like.

## labeled links
Normal links can be annotated using `[label](link)`. Internal links don't support anything like that at the moment, which makes linking with IDs cumbersome. I'm not sure what could work well with other programs that use Markdown, but here are some ideas how this could be implemented:

- In Dendron, internal links can be labeled with `[[label|link]]` (see [Dendron wiki here](https://wiki.dendron.so/notes/3472226a-ff3c-432d-bf5d-10926f39f6c2.html)). This is convenient, but I suspect it might not work well with most Markdown styles. Also discussed in [this Reddit thread](https://www.reddit.com/r/Zettlr/comments/gpozku/why_doesnt_zettlr_support_linktext_format_for/)
- `[[link]](label)` or `(label)[[link]]` would be similar to the implementation in normal links and the extra `(label)` could easily be ignored by other Markdown styles that don't support this

## better link rendering
Especially when linking with IDs, links take a lot of space without providing useful information. Here are some related ideas, although most of these would probably not be necessary anymore if labeled links were a thing:

- ID links could be collapsed to some icon or symbol unless clicked/hovered, similar to what was shown in [this Reddit post](https://www.reddit.com/r/Zettlr/comments/ohhxv0/how_to_use_zettlr_as_a_zettelkasten/) using CSS
- more options to style internal links, including the brackets `[[ ]]` could be given to the user. Maybe this is already a thing, but I didn't find a way to e.g. render the brackets smaller using custom CSS
- when adding the filename to an ID link, it could be auto-formatted, e.g. bold or in italics, to make it stand out. This is basically a cheap workaround of labeled links

## linking to note sections
Currently, it is only possible to link to a note as a whole. Even when a second ID is used in a file, or the idea is placed in the middle of a file, we can basically only link to the head of the file. See also #1800 and [this Reddit thread](https://www.reddit.com/r/Zettlr/comments/fpbjnc/is_there_a_way_to_link_to_a_specific_header_in_a/)

- [Dendron](https://www.reddit.com/r/Zettlr/comments/ohhxv0/how_to_use_zettlr_as_a_zettelkasten/) allows linking to sections with the `[[link#header]]` syntax. As above, I think this might clash with many Markdown styles, but the behavior is amazing otherwise.
- a modification of that syntax could be `[[link]]#header` or `[[link]](#header)`
- instead of allowing linking only to headers, we could have a special tag, like `#anchor<number>` to allow linking to user-defined places in any note

## transclusion
When writing notes and linking other ideas, we often need to repeat some of the content from other notes. Rendering some part of the linked note as a preview, or even a specific section would eliminate the need to repeat oneself. See also #1162 and #354

- [This Reddit thread](https://www.reddit.com/r/Zettlr/comments/nc9sbk/call_for_ideas_zettlrs_project_feature/h4p42vx?utm_source=share&utm_medium=web2x&context=3) proposes rendering internal links using the `![[link]]` syntax similar to how normal links are forced to render in-place.
- if there was a way to link to note sections, as proposed above, rendering an internal link could just include that section from the linked note
- [Roam has implemented this](https://roamstack.com/kb/starter/6-block-references/), maybe this could be used as inspiration?

## related issues and discussions
- #2048 proposes how to deal with internal links to non-existent files. While Zettlr auto-creates files when linking by title, I don't see how this can work when linking by ID
- #1925 proposes support of relative internal links like `[[path/to/file]]`, which is also discussed in [this Reddit thread](https://www.reddit.com/r/Zettlr/comments/mkmld9/use_nested_link_instead_of_finding_the_file/)
- #1798 proposes automatically appending a backlink to linked notes
- #1283 proposes a keyboard shortcut to navigate through links

**Comment by nathanlesage** [comment]
Converting this to a discussion, since I just stumbled upon it and there are many good ideas in there worth discussing!

**Comment by raphraph** [comment]
Thanks for opening a discussion on this topic.

Considering my use of Zettlr the possibility of having links to note sections would clearly be a very nice improvement. Notes can be very consistent and linking only to the note itself does not always provide enough information.

This would also allow to create links referring to another section inside the same note without having to go through the note table of contents in order to find the section which is referenced, and it would therefore allow navigating inside the note with a simple click on the note section link.

Pushing this idea there could also be a way to make links pointing to an "anchor" inside the note, which would not necessary be a section.

**Reply by nathanlesage** [reply]
> Pushing this idea there could also be a way to make links pointing to an "anchor" inside the note, which would not necessary be a section.

This is already possible. You can create a link that points to some heading-id. So if you wanted to link the heading "This is a heading!" you could create a normal Markdown link with this link contents: "#this-is-a-heading" and it will bring you there.

**Reply by raphraph** [reply]
Ah thanks for this tip !
I tried that before but I probably messed up somewhere in the markdown link syntax... I don't think there is anything in the doc though (maybe because its basic markdown ?).

I just made a few tries and it works "sort of" :)
When clicking on the link it seems to navigate to the corresponding heading, but it doesn't scroll correctly to the heading but rather "in the area" of the heading, and not always exactly at the same place. It's a "big" note which also contains tables (zettlr rendering) and images. I don't know if you already heard of this behavior (I can create an issue if needed).

The link syntax I use is  : `[test](#La reconstitution de l'écosystème)`, "La reconstitution de l'écosystème" being the heading.

Also to clarify my previous comment, I was talking about an "anchor" in order to be able to link to something which is not a heading (text, image, etc).

**Reply by nathanlesage** [reply]
So your syntax is not quite correct. The reason is that heading ids have a basic syntax that goes like this: Remove all special characters, replace spaces with hyphens and replace accents with the regular letter (except it's non-western script, of course). This is a little bit cumbersome but the idea we came up with, especially given that we don't yet support Pandoc style references. As soon as those are supported, however, it should also be possible to target and cross-reference anything other than headings. It won't come in 2.0 because there just too much work right now, but it's definitely something Zettlr should support so feel free to open a feature request for this!

Here's how Pandoc IDs work: https://pandoc.org/MANUAL.html (search for `{#` on the page)

**Reply by raphraph** [reply]
Thanks for these clarifications !

**Comment by StoltHD** [comment]
I really hope that Zettlr can support the wiki link format that foam and obsidian use with piped alias i.e. "[[Link|Alias]] the same way that Wiki actually do it, and not the other way around with the alias (or label if you want to call it that) first.
https://en.wikipedia.org/wiki/Help:Link

And if you also supported Aliases in YAML as an alternative when linking to a Note, so that if you want to have different aliases linking to the same note, you could do that by adding aliases in the YAML frontmatter of the target Note...
Reason for this is that if you work with original "sources", you want to transcribe them as accurate as possible, but different documents or a newspaper article can have small differences in the spelling, to recognize this you just add multiple possible spellings as an alias to the note and when you link to that note from another you can use that alias in the link.
And of course that we can link to either readable filenames, aliases or ID's...

I also wish that there could be support for piped links in md-tables, done the same way as i.e. obsidian or foam does it, by escaping the pipe with a backslash if it is needed to have links in a md-table.
**Reason:** when doing research, researchers often have large tables that they might want to create links from to existing Notes, i.e. if you research a set of objects you can have huge tables with Informations about those objects, example can be names of people, names of books, names of places, or names of other objects like ships, buildings, geographical locations, DNA datasets etc.
Some people might want to create links both from their existing notes for a given object, but also create "back links" from those lists/tables to their Notes of the different objects.
It is a lot easier to run a search replace in a 100k row table to create those links, than to change the format of the table to a "list" of objects with sub-lists for any property the objects might have.
And maybe a user also want to create Notes for some of those properties in a table, i.e. if you have a building, you might also have a list of owner, the location, alternative names, some dates etc. that you might want to create "sub-notes" for so creating a wiki-link to those would be of great benefit for cross-linking information...

EDIT: just added a few word to clarify in the second paragraph, and added a reason in the last.

**Comment by kyaso** [comment]
Zettlr already kind of supports naming ZKN links (see [docs](https://docs.zettlr.com/en/academic/zkn-method/#internal-linking)). As mentioned in the docs, we can't directly `Cmd/Ctrl + click` the rendered link though; we first have to click the link to reveal the embedded ZKN link, and then we can `Cmd/Ctrl + click` on the ZKN link.

There has already been an issue (#1406) that requested to make these "ZKN link embedded in Markdown link" clickable (without having to enter the link "edit mode" first).

@nathanlesage Do you still have the same view on this as in #1406?

**Reply by nathanlesage** [reply]
Mh. I still think that this is utterly hacky. But I mean if too many people really want this to be implemented, we could add an additional branch to check if the complete link text satisfies the requirement for an internal link and treat it as an internal link click. Should be fairly easy to implement.

**Comment by RoyiAvital** [comment]
I really like the embedding using `![[link]]` feature.  
For me, it is like a function in programming, write once use anywhere.

Is it something on the roadmap?

**Comment by termi-official** [comment]
Does something speak against providing some configuration option on how to render out internal links? For example, I use id's for my internal links/filenames, but rendering the id's is usually not so helpful. Instead, I would like to opt for rendering the title in place (which is provided in the YAML Frontmatter of the file with the corresponding ID).

**Comment by ThaddaeusWiedemer** [comment]
I want to chime in again with some comments on labelled links (as I called them above) or links with an alias. It seems to me that most other markdown based knowledge management/note taking tools that I'm aware of (e.g. Dendron, Obsidian, Roam) support the "wiki link" format for internal links with an alias, i.e., `[[interal_link|alias]]`. Out of all the suggestions for improving links in Zettlr, this feature seems to be the most requested one.

I'm trying to collect pros/cons of this being implemented in Zettlr to maybe make some progress on this:

**Pros**
- compatibility with other markdown-based editors (e.g. Obsidian, Dendron, Roam)
- widely used format for internal links, e.g. in Wikipedia
- synergies with another important feature of Zettlr: unique IDs that can differ from the title and can be linked to. Currently, if linking to a Zettl by ID, the link displays the ID and the actual title is appended as normal text. Links with alias would provider a cleaner way of typesetting these links
- currently only possible in a rather hacky way that does not follow the normal syntax for internal links

**Cons**
- different style of providing an alias compared to normal links, which use `[alias](link)`
- probably not supported in Pandoc and would need to be handled separately (not sure about this)

In my opinion, the different link style is not a problem, since internal links with `[[ ]]` already follow a different style than url links. Pandoc support is definitely important, but internal links are already handled separately (since they can be removed in exports), I think this wouldn't really introduce any overhead.

@nathanlesage what are your thoughts on this? Did I overlook some arguments against this? Could a compromise be for Zettlr to support this, but not making it the default?

**Reply by mitteneer** [reply]
I know it's been some time, and that there are mixed opinions about this, but did anything come of this? I think this may be the one feature I'd really like that Zettlr doesn't have...

**Reply by nathanlesage** [reply]
It's on my radar; Pandoc started to support it this year, I just hadn't have the time to implement it yet.

**Reply by mitteneer** [reply]
Cool! Thank you for the response and for all you do :)

**Comment by talestomaz** [comment]
I'm still getting used to how GItHub communities work (I'm a user, no idea of programming), so feel free to delete my comment if not appropriate.

I'd like only to show strong support for any initiative of creating "anchors" and "links" for parts or sections of notes. The lack of this functionality is hampering my progress in personal knowledge management.

Rendering parts of note would also be great for me.

Alias and labels would be an improvement as well, although I'd still prioritise internal anchoring/linking.

Zettlr is very promising and although I'm learning how to use it, it already seems to be a good improvement in my personal knowledge management. As soon as I get really comfortable with Zettlr and, hopefully, get the anchoring/linking feature, I will start to offer workshops at my university department for other faculty staff and students to train them into the combo Zettlr/Zotero. Thanks!

**Comment by IgnacioHeredia** [comment]
@ThaddaeusWiedemer 
> more options to style internal links, including the brackets [[ ]] could be given to the user. Maybe this is already a thing, but I didn't find a way to e.g. render the brackets smaller using custom CSS

You can format the brakets `[[` with CSS modifying the class `.cm-zkn-link-formatting`.

```css
body.dark #editor .CodeMirror .cm-zkn-link-formatting {
      font-size: 0.5em;
}
```

Output:
![image](https://user-images.githubusercontent.com/16413205/186979163-0d664d0e-0049-4620-863d-0ec46fe11ef3.png)

**Comment by IgnacioHeredia** [comment]
Alternatively, instead of having the ability to link to sections (or in addition to) maybe it would be easier if the internal links to IDs would take you not only to that file, but to _the specific point_ in the file where the ID is located.

I feel that this would make my Zettlekasten workflow much easier, because I usually break/merge Zettels or change section names or filenames.

In this way, one would just need to place the ID next to the text I want to reference, and not matter if I change the name of the section or the name of the file or if I move the text to another file, I will be taken always to that text.

Edit: Just opened a [new issue](https://github.com/Zettlr/Zettlr/issues/3727) with this.

**Comment by Myarlathotep** [comment]
I'm particularly interested in the subtopic included here **linking to note sections**, and I cannot see that it has been addresses anywhere explicitly. Is that correct, or have I missed something? 

This is a killer feature I think and would be greatly appreciated. 

Not just linking to headings, but it would be great it if could link to arbitrary places in files. As an example of how it might look, to explain the point, I have a list of tags like so in File A, and I want to link in File B to a specific item (via it's ID) in File A

**File A** (20240214164025.md):
- 20240214164026 Item 1
- 20240214164027 Item 2
- 20240214164028 Item 3

**File B**:
[[20240214164025]][[20240214164027]]

Which would link File B to File A, and center the screen on the ID ending 027

**Reply by Myarlathotep** [reply]
@nathanlesage Tagging because I do not know how GitHub notifications work for this sort of thing, and since this is an old discussion I guess it isn't top of the discussions page!

**Reply by Myarlathotep** [reply]
Bumping again because restrictions on my account meant these messages weren't visible apparently. Hoping you can see them now

**Reply by ElDuderino321** [reply]
Commenting on this because I was about to suggest just that! Also, I want to point out it's kinda possible already to do that, as I've been using additional IDs throughout my notes, just linking to them as if they were existing notes. If you enable the search function with CTRL-click, you will be able to access the section through the search results. Kind of a hack, but it would be nice to see this implemented in some fashion!

PS: In fact, you don't even need an ID, you can just the bit of text you want to link to, and insert it as if it were an ID, and the search function will provide it. In addition, the text will show in the pop-up bubble.
